From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dana Fried <dfried@chromium.org>
Date: Mon, 27 Jul 2026 11:54:50 -0700
Subject: [M144] [Menus, Views] Avoid potential UAF due to self-deletion.

Original change's description:
> [Menus, Views] Avoid potential UAF due to self-deletion.
>
> Bug: 536067175
> Change-Id: I107f88d2fc29a7cb5f936936acf3698b553b3af1
> Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/8127481
> Reviewed-by: Dmytro Yeroshkin <yeroshkin@chromium.org>
> Commit-Queue: Dmytro Yeroshkin <yeroshkin@chromium.org>
> Auto-Submit: Dana Fried <dfried@chromium.org>
> Commit-Queue: Dana Fried <dfried@chromium.org>
> Cr-Commit-Position: refs/heads/main@{#1666289}

(cherry picked from commit 355923849eca301965e0ce42a54d11308cd964a1)

Bug: 539349642,536067175
Change-Id: I107f88d2fc29a7cb5f936936acf3698b553b3af1
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/8157099
Auto-Submit: chrome-cherry-picker@chops-service-accounts.iam.gserviceaccount.com <chrome-cherry-picker@chops-service-accounts.iam.gserviceaccount.com>
Commit-Queue: rubber-stamper@appspot.gserviceaccount.com <rubber-stamper@appspot.gserviceaccount.com>
Bot-Commit: rubber-stamper@appspot.gserviceaccount.com <rubber-stamper@appspot.gserviceaccount.com>
Cr-Commit-Position: refs/branch-heads/7559@{#5153}
Cr-Branched-From: 223dfbac1c7542a06b422390d954afe5b560b607-refs/heads/main@{#1552494}

diff --git a/ui/views/controls/menu/menu_item_view.cc b/ui/views/controls/menu/menu_item_view.cc
index 25653e1b0c980f2650aaf51e4da9420b7fc202f6..dd9d6caafdd4bded9d5aeef0d3d19421cf3c23e6 100644
--- a/ui/views/controls/menu/menu_item_view.cc
+++ b/ui/views/controls/menu/menu_item_view.cc
@@ -63,6 +63,7 @@
 #include "ui/views/style/typography_provider.h"
 #include "ui/views/vector_icons.h"
 #include "ui/views/view_class_properties.h"
+#include "ui/views/view_tracker.h"
 #include "ui/views/view_utils.h"
 #include "ui/views/widget/widget.h"
 
@@ -1028,20 +1029,30 @@ void MenuItemView::UpdateEmptyMenusAndMetrics() {
   // visible items. Copy the children, since we may mutate them as we go.
   const Views children = submenu_->children();
   bool has_visible_menu_items = false;
+
+  // Making changes to selection, etc. can cause `this` to be deleted. Track the
+  // continued existence of this object while updating.
+  ViewTracker tracker(this);
+
   for (View* child : children) {
     MenuItemView* const child_menu = AsViewClass<MenuItemView>(child);
     if (!child_menu) {
       continue;
     }
     if (IsViewClass<EmptyMenuMenuItem>(child)) {
+      const bool selected = child_menu->IsSelected();
       // Prevent view destruction until selection is updated.
       // We remove the child before updating selection in case of re-entrancy.
       std::unique_ptr<View> removed_child = submenu_->RemoveChildViewT(child);
-      if (child_menu->IsSelected()) {
+      if (tracker && selected) {
         // Update selection to this menu before deleting the currently
         // selected child.
         GetMenuController()->SetSelection(
             this, MenuController::SELECTION_UPDATE_IMMEDIATELY);
+        // This can also delete `this`, so check for that.
+        if (!tracker) {
+          return;
+        }
       }
       submenu_
           ->InvalidateLayout();  // Ideally the submenu would have a layout
